transformer_engine/jax/cpp_extensions/attention.py

```diff
@@ -1057,7 +1057,7 @@ def partition(config, mesh, arg_infos, result_infos):
         # Call base implementation for non-context parallel mesh to avoid unecessary work.
         is_context_parallel = get_mesh_axis_size(config.cp_axis, mesh) > 1
         assert (
-            is_context_parallel and config.window_size[0] == -1
+            not is_context_parallel or config.window_size[0] == -1
```

Comment:
I'm not sure I understand this change. This assertion is meant to catch the case when sliding window attention is being used with Context parallelism, so the existing check is correct. Could you post some more details (and a stacktrace if possible) about your bug?

Reply:
Please have a look at this: https://github.com/NVIDIA/TransformerEngine/pull/1205#discussion_r1776149416

Reply:
Hi @ksivaman , the current assert logic is becoming false for all non CP related runs, hence blocking them. I am running without CP and without SWA but still seeing the `AssertionError`:
```
...
number parameters: 6.738 billion
Per train step:
 Total TFLOPs: 755.06 
 split as 86.02% learnable weight flops and 13.98% attention flops
jax.errors.SimplifiedTraceback: For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/maxtext/MaxText/train.py", line 776, in <module>
    app.run(main)
  File "/usr/local/lib/python3.10/dist-packages/absl/app.py", line 308, in run
    _run_main(main, args)
  File "/usr/local/lib/python3.10/dist-packages/absl/app.py", line 254, in _run_main
    sys.exit(main(argv))
  File "/opt/maxtext/MaxText/train.py", line 772, in main
    train_loop(config)
  File "/opt/maxtext/MaxText/train.py", line 666, in train_loop
    state, metrics = p_train_step(state, example_batch, nextrng)
jaxlib.xla_extension.XlaRuntimeError: INTERNAL: custom_partitioner: Traceback (most recent call last):
  File "/opt/maxtext/MaxText/train.py", line 776, in <module>
  File "/usr/local/lib/python3.10/dist-packages/absl/app.py", line 308, in run
  File "/usr/local/lib/python3.10/dist-packages/absl/app.py", line 254, in _run_main
  File "/opt/maxtext/MaxText/train.py", line 772, in main
  File "/opt/maxtext/MaxText/train.py", line 666, in train_loop
  File "/opt/jax/jax/_src/traceback_util.py", line 180, in reraise_with_filtered_traceback
  File "/opt/jax/jax/_src/pjit.py", line 338, in cache_miss
  File "/opt/jax/jax/_src/pjit.py", line 188, in _python_pjit_helper
  File "/opt/jax/jax/_src/core.py", line 2781, in bind
  File "/opt/jax/jax/_src/core.py", line 442, in bind_with_trace
  File "/opt/jax/jax/_src/core.py", line 948, in process_primitive
  File "/opt/jax/jax/_src/pjit.py", line 1737, in _pjit_call_impl
  File "/opt/jax/jax/_src/pjit.py", line 1713, in call_impl_cache_miss
  File "/opt/jax/jax/_src/pjit.py", line 1643, in _pjit_call_impl_python
  File "/opt/jax/jax/_src/interpreters/pxla.py", line 2321, in compile
  File "/opt/jax/jax/_src/interpreters/pxla.py", line 2835, in from_hlo
  File "/opt/jax/jax/_src/interpreters/pxla.py", line 2647, in _cached_compilation
  File "/opt/jax/jax/_src/compiler.py", line 427, in compile_or_get_cached
  File "/opt/jax/jax/_src/compiler.py", line 655, in _compile_and_write_cache
  File "/opt/jax/jax/_src/profiler.py", line 333, in wrapper
  File "/opt/jax/jax/_src/compiler.py", line 267, in backend_compile
  File "/opt/jax/jax/_src/custom_partitioning.py", line 155, in _custom_partitioning_partition
  File "/opt/transformer-engine/transformer_engine/jax/cpp_extensions/attention.py", line 1060, in partition
AssertionError: Sliding window attention is not supported when context parallelism is enabled
```

Reply:
It seems that the changes suggested by @zlsh80826 here https://github.com/NVIDIA/TransformerEngine/pull/1205#discussion_r1791106333 does not work. Probably this one should be the correct:
```python
is_context_parallel = get_mesh_axis_size(config.cp_axis, mesh) > 1
if is_context_parallel and config.window_size[0] > 0:
    assert (
        is_context_parallel and config.window_size[0] == -1
    ), "Sliding window attention is not supported when context parallelism is enabled"
```
So when `is_context_parallel  = False`, the assert will not be triggered.

Reply:
`not is_context_parallel or config.window_size[0] == -1` should be correct.

Reply:
so, the PR should be good, right? @huanghua1994

Reply:
Yes, I think the PR is good. Should be fine to merge after testing.

Reply:
Oh yes, this is correct.